pkg/lca/ibu_test.go

```diff
@@ -0,0 +1,753 @@
+package lca
+
+import (
+	"fmt"
+	"testing"
+
+	lcav1alpha1 "github.com/openshift-kni/lifecycle-agent/api/v1alpha1"
+	"github.com/stretchr/testify/assert"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+
+	"github.com/openshift-kni/eco-goinfra/pkg/clients"
+)
+
+var (
+	ibulGvk = schema.GroupVersionKind{
+		Group:   "lca.openshift.io",
+		Version: "v1alpha1",
+		Kind:    "ImageBasedUpgrade",
+	}
+)
+
+func TestImageBasedUpgradeWithOptions(t *testing.T) {
+	testSettings := buildTestClientWithDummyObject()
+	testBuilder, _ := PullImageBasedUpgrade(testSettings)
+	testBuilder = testBuilder.WithOptions(
+		func(builder *ImageBasedUpgradeBuilder) (*ImageBasedUpgradeBuilder, error) {
+			return builder, nil
+		})
+
+	assert.Equal(t, "", testBuilder.errorMsg)
+
+	testBuilder = testBuilder.WithOptions(
+		func(builder *ImageBasedUpgradeBuilder) (*ImageBasedUpgradeBuilder, error) {
+			return builder, fmt.Errorf("error")
+		})
+	assert.Equal(t, "error", testBuilder.errorMsg)
+}
+
+func buildTestClientWithDummyObject() *clients.Settings {
+	return clients.GetTestClients(clients.TestClientParams{
+		K8sMockObjects: buildDummyIBU(),
+		GVK:            []schema.GroupVersionKind{ibulGvk},
+	})
+}
+
+func buildDummyIBU() []runtime.Object {
+	return append([]runtime.Object{}, &lcav1alpha1.ImageBasedUpgrade{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "upgrade",
+		},
+		Spec: lcav1alpha1.ImageBasedUpgradeSpec{
+			Stage: "Idle",
+		},
+	})
+}
+
+func TestImageBasedUpgradePull(t *testing.T) {
+	testCases := []struct {
+		expectedError       bool
+		addToRuntimeObjects bool
+		expectedErrorText   string
+	}{
+		{
+			expectedError:       false,
+			addToRuntimeObjects: true,
+		},
+	}
+
+	for _, testCase := range testCases {
+		var (
+			runtimeObjects []runtime.Object
+			testSettings   *clients.Settings
+		)
+
+		testIBU := generateImageBasedUpgrade()
+
+		if testCase.addToRuntimeObjects {
+			runtimeObjects = append(runtimeObjects, testIBU)
+		}
+
+		testSettings = clients.GetTestClients(clients.TestClientParams{
+			K8sMockObjects: runtimeObjects,
+		})
+
+		// Test the PullImageBasedUpgrade function
+		builderResult, err := PullImageBasedUpgrade(testSettings)
+
+		// Check the error
+		if testCase.expectedError {
+			assert.NotNil(t, err)
+
+			if testCase.expectedErrorText != "" {
+				assert.Equal(t, testCase.expectedErrorText, err.Error())
+			}
+		} else {
+			assert.Nil(t, err)
+			assert.NotNil(t, builderResult)
+		}
+	}
+}
+
+func TestImageBasedUpgradeUpdate(t *testing.T) {
```

Comment:
Please check reference below:

```
func TestArgoCdUpdate(t *testing.T) {
	testCases := []struct {
		testArgoCd    *Builder
		expectedError error
		image         string
	}{
		{
			testArgoCd:    buildValidArgoCdBuilder(buildArgoCdTestClientWithDummyObject()),
			expectedError: nil,
			image:         "testimage",
		},
		{
			testArgoCd:    buildInValidArgoCdBuilder(buildArgoCdTestClientWithDummyObject()),
			expectedError: fmt.Errorf("argocd 'nsname' cannot be empty"),
			image:         "testimage",
		},
		{
			testArgoCd:    buildValidArgoCdBuilder(clients.GetTestClients(clients.TestClientParams{})),
			expectedError: fmt.Errorf("argocds.argoproj.io \"argocd\" not found"),
		},
	}

	for _, testCase := range testCases {
		assert.Equal(t, "", testCase.testArgoCd.Definition.Spec.Image)
		testCase.testArgoCd.Definition.ResourceVersion = "999"
		assert.Equal(t, "", testCase.testArgoCd.Definition.Spec.Image)
		testCase.testArgoCd.Definition.Spec.Image = testCase.image
		_, err := testCase.testArgoCd.Update(false)

		if errors.IsNotFound(err) {
			assert.Equal(t, testCase.expectedError.Error(), err.Error())
		} else {
			assert.Equal(t, testCase.expectedError, err)
		}

		if testCase.expectedError == nil {
			assert.Equal(t, testCase.image, testCase.testArgoCd.Object.Spec.Image)
			assert.Equal(t, testCase.testArgoCd.Object, testCase.testArgoCd.Definition)
		}
	}
}
```

Reply:
I don't see how its a good reference.
testArgoCd is a pointer to Builder in the struct.

to set testArgoCd we then call  buildValidArgoCdBuilder which calls the NewBuilder function.
lca doesn't have this function